examples/jsm/renderers/webgpu/WebGPURenderer.js

```diff
@@ -0,0 +1,601 @@
+
+		if ( x === null ) {
+
+			this._viewport = null;
+
+		} else {
+
+			this._viewport = {
+				x: x,
+				y: y,
+				width: width,
+				height: height,
+				minDepth: minDepth,
+				maxDepth: maxDepth
+			};
+
+		}
+
+	}
+
+	dispose() {
+
+		this._objects.dispose();
+		this._properties.dispose();
+		this._renderPipelines.dispose();
+		this._bindings.dispose();
+		this._info.dispose();
+		this._renderLists.dispose();
+
+	}
+
+	_projectObject( object, camera, groupOrder ) {
+
+		const info = this._info;
+		const currentRenderList = this._currentRenderList;
+
+		if ( object.visible === false ) return;
+
+		const visible = object.layers.test( camera.layers );
+
+		if ( visible ) {
+
+			if ( object.isGroup ) {
+
+				groupOrder = object.renderOrder;
+
+			} else if ( object.isLOD ) {
+
+				if ( object.autoUpdate === true ) object.update( camera );
+
+			} else if ( object.isLight ) {
+
+				//currentRenderState.pushLight( object );
+
+				if ( object.castShadow ) {
+
+					//currentRenderState.pushShadow( object );
+
+				}
+
+			} else if ( object.isSprite ) {
+
+				if ( ! object.frustumCulled || _frustum.intersectsSprite( object ) ) {
+
+					if ( this.sortObjects === true ) {
+
+						_vector3.setFromMatrixPosition( object.matrixWorld ).applyMatrix4( _projScreenMatrix );
+
+					}
+
+					const geometry = object.geometry;
+					const material = object.material;
+
+					if ( material.visible ) {
+
+						currentRenderList.push( object, geometry, material, groupOrder, _vector3.z, null );
+
+					}
+
+				}
+
+			} else if ( object.isMesh || object.isLine || object.isPoints ) {
+
+				if ( object.isSkinnedMesh ) {
+
+					// update skeleton only once in a frame
+
+					if ( object.skeleton.frame !== info.render.frame ) {
+
+						object.skeleton.update();
+						object.skeleton.frame = info.render.frame;
+
+					}
+
+				}
+
+				if ( ! object.frustumCulled || _frustum.intersectsObject( object ) ) {
+
+					if ( this.sortObjects === true ) {
+
+						_vector3.setFromMatrixPosition( object.matrixWorld ).applyMatrix4( _projScreenMatrix );
+
+					}
+
+					const geometry = object.geometry;
+					const material = object.material;
+
+					if ( Array.isArray( material ) ) {
+
+						const groups = geometry.groups;
+
+						for ( let i = 0, l = groups.length; i < l; i ++ ) {
+
+							const group = groups[ i ];
+							const groupMaterial = material[ group.materialIndex ];
+
+							if ( groupMaterial && groupMaterial.visible ) {
+
+								currentRenderList.push( object, geometry, groupMaterial, groupOrder, _vector3.z, group );
+
+							}
+
+						}
+
+					} else if ( material.visible ) {
+
+						currentRenderList.push( object, geometry, material, groupOrder, _vector3.z, null );
+
+					}
+
+				}
+
+			}
+
+		}
+
+		const children = object.children;
+
+		for ( let i = 0, l = children.length; i < l; i ++ ) {
+
+			this._projectObject( children[ i ], camera, groupOrder );
+
+		}
+
+	}
```

Comment:
`projectObject()` is not WebGPU/WebGL specific logic. So ideally it would be good to export at some point to external class and make it resuable in `WebGPURenderer` and `WebGLRenderer`?

Reply:
Sometimes it's okay to duplicate code.

Reply:
Yes, a few components could be actually shared across both renderers. When the renderer gets more stable and we still have more or less duplicated code sections, I vote for encapsulate it in some way.

Reply:
> When the renderer gets more stable and we still have more or less duplicated code sections, we should reconsider this.

Agreed. So I wrote "at some point".